### ISSUE TITLE
CCD collision + collision group = problems ?

### ISSUE THREAD

**Issue by ekureuil** [jMonkeyEngine/jmonkeyengine#1283]
Below is a small code showing the problem.

I used the TestCDD example and modified it a little bit to include collision groups.

As a refresher, from the source documentation (setCollisionGroup):

>      Groups are represented by integer bit masks with exactly 1 bit set.
>      Pre-made variables are available in PhysicsCollisionObject. By default, physics objects are in COLLISION_GROUP_01.
>      Two objects can collide only if one of them has the collisionGroup of the other in its collideWithGroups set.
> 

In my example we have:
 - a wall and a ground that are in group 1 and collide with objects of group 2
 - a ball (made by makeBallCCD1): no collision group specified, so ball is in group 1 and collide with group 1
 - another ball (made by makeBallCCD2): collision group is 3, and collide with group 4 (nothing in that group)
 - both ball use CCD

what i would expect:
 - wall/ground should not collide with any ball (ball1 and ball2 are not in collision group 2)
 - ball1 should collide with wall/ground (wall/ground are in collision group 1) (how does it work ? which is taken into account ?)
 - ball2 should definitely not collide with wall/ground (nor with ball1)

What i see:
 - ball2 collide with wall/ground but do not roll as ball1 does ...
 - if you shoot a ball1 so that the ball hit the wall straight ahead and without much velocity you can throw a ball2 at the first ball1 and the ball2 will collide with ball1 ...
 - after 2 or 3 rebound, ball2 will go trough the ground as if it remembered it was not supposed to collide !
 - ball1 collide happily with ground and wall (as intended ?)

> FILE TestCCD.java:

>     import com.jme3.app.SimpleApplication;
>     import com.jme3.bullet.BulletAppState;
>     import com.jme3.input.MouseInput;
>     import com.jme3.input.controls.ActionListener;
>     import com.jme3.input.controls.MouseButtonTrigger;
>     import com.jme3.scene.Geometry;
>     
>      public class TestCCD extends SimpleApplication implements ActionListener {
>         BulletAppState bas;
>         @Override
>         public void simpleInitApp() {
>             bas = new BulletAppState();
>             bas.setDebugEnabled(true);
>     
>            Tools.initGlobals(rootNode, bas, assetManager, getCamera());
>             stateManager.attach(bas);
>     
>             Tools.makeWall();
>             Tools.makeGround();
>     
>             flyCam.setEnabled(true);
>             flyCam.setMoveSpeed(40f);
>     
>             inputManager.addMapping("shoot", new MouseButtonTrigger(MouseInput.BUTTON_LEFT));
>             inputManager.addMapping("shoot2", new MouseButtonTrigger(MouseInput.BUTTON_RIGHT));
>             inputManager.addListener(this, "shoot");
>             inputManager.addListener(this, "shoot2");
>     
>         }
>     
>         @Override
>         public void onAction(String binding, boolean value, float tpf) {
>             if (binding.equals("shoot") && !value) {
>                 Tools.makeBallCCD1();
>             } else if (binding.equals("shoot2") && !value) {
>                 Tools.makeBallCCD2();
>             }
>         }
>     
>         public static void main(String[] args) {
>             TestCCD app = new TestCCD();
>             app.start();
>         }
>     }

> File: Tools.java

>     import com.jme3.asset.AssetManager;
>     import com.jme3.bullet.BulletAppState;
>     import com.jme3.bullet.collision.PhysicsCollisionObject;
>     import com.jme3.bullet.collision.shapes.MeshCollisionShape;
>     import com.jme3.bullet.collision.shapes.SphereCollisionShape;
>     import com.jme3.bullet.control.RigidBodyControl;
>     import com.jme3.material.Material;
>     import com.jme3.math.ColorRGBA;
>     import com.jme3.math.Vector3f;
>     import com.jme3.renderer.Camera;
>     import com.jme3.renderer.queue.RenderQueue;
>     import com.jme3.scene.Geometry;
>     import com.jme3.scene.Node;
>     import com.jme3.scene.shape.Box;
>     import com.jme3.scene.shape.Sphere;
>     
>     public class Tools {
>     
>         static Node rootNode;
>         static BulletAppState bullet;
>         static AssetManager am;
>         static Sphere ball;
>         static SphereCollisionShape ballCS;
>         static Camera cam;
>     
>         static void initGlobals(Node root, BulletAppState bas, AssetManager am, Camera camera) {
>             rootNode = root;
>             bullet = bas;
>             Tools.am = am;
>             cam = camera;
>     
>     
>             ball = new Sphere(32, 32, 0.4f, true, false);
>             ball.setTextureMode(Sphere.TextureMode.Projected);
>             ballCS = new SphereCollisionShape(0.5f);
>     
>         }
>     
>         static Node createCollisionBox(String name, Vector3f position, Vector3f extend) {
>             Node box = new Node();
>             box.setLocalTranslation(position);
>     
>             RigidBodyControl boxRBC = new RigidBodyControl(new MeshCollisionShape(new Box(extend.x, extend.y, extend.z)), 0f);
>             boxRBC.setCollisionGroup(PhysicsCollisionObject.COLLISION_GROUP_01);
>             boxRBC.setCollideWithGroups(PhysicsCollisionObject.COLLISION_GROUP_02);
>     
>             box.addControl(boxRBC);
>     
>             rootNode.attachChild(box);
>             bullet.getPhysicsSpace().add(box);
>     
>             return box;
>         }
>     
>         static Node makeWall() {
>             return createCollisionBox("Wall", new Vector3f(2.5f, 0f, 0f), new Vector3f(4f, 5f, 0.0f));
>         }
>     
>         static Node makeGround() {
>             return createCollisionBox("Ground", new Vector3f(0f, -6f, 0f), new Vector3f(100f, 1f, 100f));
>         }
>     
>         static void makeBallCCD1() {
>             Geometry ballGeom = new Geometry("ball1", ball);
>     
>             Material mat = new Material(am, "Common/MatDefs/Misc/Unshaded.j3md");
>             mat.getAdditionalRenderState().setWireframe(true);
>             mat.setColor("Color", ColorRGBA.Green);
>     
>             ballGeom.setMaterial(mat);
>             ballGeom.setName("ball1");
>             ballGeom.setLocalTranslation(cam.getLocation());
>             ballGeom.setShadowMode(RenderQueue.ShadowMode.CastAndReceive);
>     
>             ballGeom.addControl(new RigidBodyControl(ballCS, 1));
>             ballGeom.getControl(RigidBodyControl.class).setCcdMotionThreshold(0.01f);
>             ballGeom.getControl(RigidBodyControl.class).setLinearVelocity(cam.getDirection().mult(40));
>             rootNode.attachChild(ballGeom);
>             bullet.getPhysicsSpace().add(ballGeom);
>         }
>     
>         static void makeBallCCD2() {
>             Geometry ballGeom = new Geometry("ball2", ball);
>     
>             Material mat2 = new Material(am, "Common/MatDefs/Misc/Unshaded.j3md");
>             mat2.getAdditionalRenderState().setWireframe(true);
>             mat2.setColor("Color", ColorRGBA.Red);
>     
>             ballGeom.setMaterial(mat2);
>             ballGeom.setName("ball2");
>             ballGeom.setLocalTranslation(cam.getLocation());
>             ballGeom.setShadowMode(RenderQueue.ShadowMode.CastAndReceive);
>     
>             RigidBodyControl rbc = new RigidBodyControl(ballCS, 1);
>             rbc.setCollisionGroup(PhysicsCollisionObject.COLLISION_GROUP_03);
>             rbc.setCollideWithGroups(PhysicsCollisionObject.COLLISION_GROUP_04);
>     
>             ballGeom.addControl(rbc);
>     
>             ballGeom.getControl(RigidBodyControl.class).setCcdMotionThreshold(0.01f);
>     
>             ballGeom.getControl(RigidBodyControl.class).setLinearVelocity(cam.getDirection().mult(40));
>             rootNode.attachChild(ballGeom);
>             bullet.getPhysicsSpace().add(ballGeom);
>         }
>     }

On an unrelated note, i could not attach a java file to this issue (only pictures ???) and the code block was not working properly (at least in the preview). Instead i used quotation block (which is tedious). This is my first time opening an issue, did i do something wrong ? This seems counter-intuitive to not be able to attach code files ...

**Comment by stephengold** [comment]
Thank you for your bug report.

Before I attempt to reproduce the issue, I'd like to know which physics library and version you were using. For instance, "jme3-jbullet 3.2.4-stable" or "jme3-bullet 3.3.0-beta1". There's a big difference between Bullet and JBullet.

To include multi-line code samples in a GitHub issue, I usually put a triple-backtick before and after the sample.

For troubleshooting help, I recommend posting to the JME Forum (https://hub.jmonkeyengine.org/) which should attract more attention than GitHub.

**Comment by ekureuil** [comment]
Oops, sorry, that was with jme3-bullet-3.3.0-beta1, i do not use JBullet.

With 3.2.4-stable, i get a different behavior:

Ball2 does not seem to collide much with ground/wall, except that sometimes it will collide with ground/wall in a weird way (like if there was a vertical invisible wall (maybe at the normal of the contact point ?))... Also it still collides with Ball1 which it should not do.

**Comment by stephengold** [comment]
The discussion continued at
  https://hub.jmonkeyengine.org/t/ccd-collision-collision-group-problems/42806

**Comment by stephengold** [comment]
I'm investigating this issue. I've discovered that enabling CCD prevents `jmeFilterCallback::needBroadphaseCollision()` from being invoked, but I don't understand the mechanism yet.

**Comment by stephengold** [comment]
I suspected PR #1031 might be involved, but that hypothesis is ruled out because this issue occurs in jme3-bullet v3.2.0, which didn't include PR #1031.

Interestingly, this issue does NOT occur in jme3-bullet v3.1.0. One thing that changed between v3.1-stable and v3.2-stable was the Bullet version, which jumped from 2.82-r2704 to 2.86.1

**Comment by stephengold** [comment]
I've been studying Bullet (C++) code with the help of GDB. Predictive contacts (generated by CCD) are separate from the overlapping-pair cache (which is where JME's collision-group filtering takes place).

Bullet has a filtering mechanism for predictive contacts, but I don't see any way to override it. Here's where filtering of predictive contacts takes place:

```
#0  in btCollisionWorld::ConvexResultCallback::needsCollision(btBroadphaseProxy*) const 
    at BulletCollision/CollisionDispatch/btCollisionWorld.h:355
#1  in btClosestNotMeConvexResultCallback::needsCollision(btBroadphaseProxy*) const 
    at BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp:801
#2  in btSingleSweepCallback::process(btBroadphaseProxy const*) 
    at BulletCollision/CollisionDispatch/btCollisionWorld.cpp:1061
#3  in BroadphaseRayTester::Process(btDbvtNode const*) 
    at BulletCollision/BroadphaseCollision/btDbvtBroadphase.cpp:236
#4  in btDbvt::rayTestInternal(btDbvtNode const*, btVector3 const&, btVector3 const&, btVector3 const&, unsigned int*, float, btVector3 const&, btVector3 const&, btAlignedObjectArray<btDbvtNode const*>&, btDbvt::ICollide&) const
    at BulletCollision/BroadphaseCollision/btDbvt.h:1267
#5  in btDbvtBroadphase::rayTest(btVector3 const&, btVector3 const&, btBroadphaseRayCallback&, btVector3 const&, btVector3 const&) 
    at BulletCollision/BroadphaseCollision/btDbvtBroadphase.cpp:263
#6  in btCollisionWorld::convexSweepTest(btConvexShape const*, btTransform const&, btTransform const&, btCollisionWorld::ConvexResultCallback&, float) const 
    at BulletCollision/CollisionDispatch/btCollisionWorld.cpp:1103
#7  in btDiscreteDynamicsWorld::createPredictiveContactsInternal(btRigidBody**, int, float) 
    at BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp:891
#8  in btDiscreteDynamicsWorld::createPredictiveContacts(float) 
    at BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp:939
#9  in btDiscreteDynamicsWorld::internalSingleStepSimulation(float) 
    at BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp:470
```

Perhaps a filter override (similar to `btOverlappingPairCache::setOverlapFilterCallback()`) could be added to `btDiscreteDynamicsWorld`.

Alternatively, `btClosestNotMeConvexResultCallback` contains a pointer to the pair cache. Perhaps the cache's filter could be applied somehow. 

For comparison, here's where filtering of overlapping pairs takes place:

```
#0  in jmePhysicsSpace::jmeFilterCallback::needBroadphaseCollision(btBroadphaseProxy*, btBroadphaseProxy*) const
    at jmePhysicsSpace.cpp:107
#1  in btHashedOverlappingPairCache::needsBroadphaseCollision(btBroadphaseProxy*, btBroadphaseProxy*) const
    at BulletCollision/BroadphaseCollision/btOverlappingPairCache.h:111
#2  in btHashedOverlappingPairCache::addOverlappingPair(btBroadphaseProxy*, btBroadphaseProxy*)
    at BulletCollision/BroadphaseCollision/btOverlappingPairCache.h:125
#3  in btDbvtTreeCollider::Process(btDbvtNode const*, btDbvtNode const*)
    at BulletCollision/BroadphaseCollision/btDbvtBroadphase.cpp:116
#4  in btDbvt::collideTTpersistentStack(btDbvtNode const*, btDbvtNode const*, btDbvt::ICollide&) 
    at BulletCollision/BroadphaseCollision/btDbvt.h:1070
#5 in btDbvtBroadphase::setAabb(btBroadphaseProxy*, btVector3 const&, btVector3 const&, btDispatcher*)
    at BulletCollision/BroadphaseCollision/btDbvtBroadphase.cpp:369
#6  in btCollisionWorld::updateSingleAabb(btCollisionObject*)
    at BulletCollision/CollisionDispatch/btCollisionWorld.cpp:177
#7  in btCollisionWorld::updateAabbs()
    at BulletCollision/CollisionDispatch/btCollisionWorld.cpp:210
#8  in btCollisionWorld::performDiscreteCollisionDetection()
    at BulletCollision/CollisionDispatch/btCollisionWorld.cpp:227
#9 in btDiscreteDynamicsWorld::internalSingleStepSimulation(float)
    at BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp:473
```

**Comment by stephengold** [comment]
I've submitted PR 2620 to https://github.com/bulletphysics/bullet3 for the alternative strategy.

**Comment by stephengold** [comment]
The bullet3 PR has been integrated, so now this should be a fairly easy fix.

**Comment by stephengold** [comment]
This issue isn't really fixed until GitHub Actions deploys a new natives snapshot. I expected that to happen here:  https://github.com/jMonkeyEngine/jmonkeyengine/runs/621387648
but the log says "No changes, skip."

Further investigation seems warranted.

**Comment by stephengold** [comment]
I've opened #1365 for the build issue.

**Comment by stephengold** [comment]
The natives snapshot got updated. Fixed in master branch as of f7352ffcfea415d366e3eb8cc90f2dadfd3db96c
